src/coreclr/jit/valuenum.cpp

```diff
@@ -6001,6 +6001,84 @@ void ValueNumStore::SetVNIsCheckedBound(ValueNum vn)
     m_checkedBoundVNs.AddOrUpdate(vn, true);
 }
 
+#ifdef FEATURE_HW_INTRINSICS
+ValueNum ValueNumStore::EvalHWIntrinsicFunUnary(
+    var_types type, NamedIntrinsic ni, VNFunc func, ValueNum arg0VN, bool encodeResultType, ValueNum resultTypeVN)
+{
+    if (IsVNConstant(arg0VN))
+    {
+        switch (ni)
+        {
+#ifdef TARGET_ARM64
+            case NI_ArmBase_LeadingZeroCount:
+#else
+            case NI_LZCNT_LeadingZeroCount:
+#endif
+            {
+                UINT32 cns = (UINT32)GetConstantInt32(arg0VN);
+                int    lzc = 0;
+                while (cns != 0)
+                {
+                    cns = cns >> 1;
+                    lzc++;
+                }
+                return VNForIntCon(32 - lzc);
```

Comment:
Do we need to avoid folding this for `0` input? I see this note:
"It should be noted that on processors that do not support LZCNT, the instruction byte encoding is executed as BSR."
and `BSR` on 0 is undefined, so potentially we could see different results in tier 0/tier 1? Or do we not support the instruction in those cases?

Reply:
Ah, I guess we cannot support it. I thought `BSR` and `lzcnt` were essentially the same except for on `0`, but turns out it's not